New documentation structure (#2718)
huggingface/datasets: docs/source/dataset_script.rst

Dataset script
==============

Write a dataset script to load and share your own datasets. It is a Python file that defines the different configurations and splits of your dataset, as well as how to download and process the data.

Any dataset script, for example ``my_dataset.py``, can be placed in a folder or a repository named ``my_dataset`` and be loaded with:

.. code-block::

    >>> from datasets import load_dataset
    >>> load_dataset("path/to/my_dataset")

The following guide includes instructions for dataset scripts for how to:

* Add dataset metadata.
* Download data files.
* Generate samples.
* Test if your dataset was generated correctly.
* Create a Dataset card.

> * Upload a dataset to the Hugging Face Hub or Github.

Reviewer: ```suggestion
* Upload a dataset to the Hugging Face Hub or GitHub.
```

Add dataset attributes
----------------------

The first step is to add some information, or attributes, about your dataset in :func:`datasets.DatasetBuilder._info`. The most important attributes you should specify are:

1. :obj:`datasets.DatasetInfo.description` provides a concise description of your dataset. The description informs the user what's in the dataset, how it was collected, and how it can be used for a NLP task.

2. :obj:`datasets.DatasetInfo.features` defines the name and type of each column in your dataset. This will also provide the structure for each example, so it is possible to create nested subfields in a column if you want. Take a look at :class:`datasets.Features` for a full list of feature types you can use.

.. code-block::

   datasets.Features(
       {
           "id": datasets.Value("string"),
           "title": datasets.Value("string"),
           "context": datasets.Value("string"),
           "question": datasets.Value("string"),
           "answers": datasets.Sequence(
               {
                   "text": datasets.Value("string"),
                   "answer_start": datasets.Value("int32"),
               }
           ),
       }
   )

> 3. :obj:`datasets.DatasetInfo.homepage` contains a URL to the datasets homepage so users can find more details about the dataset.

Reviewer: ```suggestion
3. :obj:`datasets.DatasetInfo.homepage` contains the URL to the dataset homepage so users can find more details about the dataset.
```

> 4. :obj:`datasets.DatasetInfo.citation` contains a BibTex citation for the dataset.

Reviewer: ```suggestion
4. :obj:`datasets.DatasetInfo.citation` contains a BibTeX citation for the dataset.
```

                   "context": datasets.Value("string"),
                   "question": datasets.Value("string"),
                   "answers": datasets.features.Sequence(
                       {"text": datasets.Value("string"), "answer_start": datasets.Value("int32"),}
                   ),
               }
           ),
           # No default supervised_keys (as we have to pass both question
           # and context as input).
           supervised_keys=None,
           homepage="https://rajpurkar.github.io/SQuAD-explorer/",
           citation=_CITATION,
       )

Multiple configurations
^^^^^^^^^^^^^^^^^^^^^^^

In some cases, your dataset may have multiple configurations. For example, the `SuperGLUE <https://huggingface.co/datasets/super_glue>`_ dataset is a collection of 5 datasets designed to evaluate language understanding tasks. 🤗 Datasets provides :class:`datasets.BuilderConfig` which allows you to create different configurations for the user to select from.

Let's study the `SuperGLUE loading script <https://github.com/huggingface/datasets/blob/master/datasets/super_glue/super_glue.py>`_ to see how you can define several configurations.

> 1. Create a :class:`datasets.BuilderConfig` class with attributes about your dataset. These attributes can be the features of your dataset, label classes, and a URL to the data files.

Reviewer: ```suggestion
1. Create a :class:`datasets.BuilderConfig` subclass with attributes about your dataset. These attributes can be the features of your dataset, label classes, and a URL to the data files.
```

       """BuilderConfig for SuperGLUE.

       Args:
       features: `list[string]`, list of the features that will appear in the
           feature dict. Should not include "label".
       data_url: `string`, url to download the zip file from.
       citation: `string`, citation for the data set.
       url: `string`, url for information about the data set.
       label_classes: `list[string]`, the list of classes for the label if the
           label is present as a string. Non-string labels will be cast to either
           'False' or 'True'.
       **kwargs: keyword arguments forwarded to super.
       """
       # Version history:
       # 1.0.2: Fixed non-nondeterminism in ReCoRD.
       # 1.0.1: Change from the pre-release trial version of SuperGLUE (v1.9) to
       #        the full release (v2.0).
       # 1.0.0: S3 (new shuffling, sharding and slicing mechanism).
       # 0.0.2: Initial version.
       super(SuperGlueConfig, self).__init__(version=datasets.Version("1.0.2"), **kwargs)
       self.features = features
       self.label_classes = label_classes
       self.data_url = data_url
       self.citation = citation
       self.url = url

> 2. Sub-class the base :class:`datasets.BuilderConfig` to add additional attributes of a configuration. This gives you more flexibility to specify the name and description of each configuration. These sub-classes should be listed under :obj:`datasets.DatasetBuilder.BUILDER_CONFIGS`:

Reviewer: This is redundant with previous item. There we already had to subclass `BuilderConfig` in order to add additional attributes, like `features`, `label_classes` and `citation`. Note that the only attributes of base `BuilderConfig` are: `name`, `version`, `data_dir`, `data_files` and `description`.

Reviewer: ```suggestion
2. Sub-class the base :class:`datasets.BuilderConfig` to add additional attributes of a configuration. This gives you more flexibility to specify all the name and description of each configuration. These sub-class instances should be listed under :obj:`datasets.DatasetBuilder.BUILDER_CONFIGS`:
```

Reviewer 2: I think this point is about creating the config instances and list them in BUILDER_CONFIGS.

Let me try to rephrase it

   ]

   DEFAULT_CONFIG_NAME = "first_domain"

.. important::

   Only use a default configuration when it makes sense. Don't set one because it may be more convenient for the user to not specify a configuration when they load your dataset. For example, multi-lingual datasets often have a separate configuration for each language. An appropriate default may be an aggregated configuration that loads all the languages of the dataset if the user doesn't request a particular one.

Download data files and organize splits
---------------------------------------

After you've defined the attributes of your dataset, the next step is to download the data files and organize them according to their splits. 

1. Create a dictionary of URLs in the loading script that point to the original SQuAD data files:

.. code-block::

   _URL = "https://rajpurkar.github.io/SQuAD-explorer/dataset/"
       _URLS = {
           "train": _URL + "train-v1.1.json",
           "dev": _URL + "dev-v1.1.json",
       }

.. tip::
    If the data files live in the same folder or repository of the dataset script, you can just pass the relative paths to the files instead of URLs.

> 2. :obj:`datasets.DownloadManager.download_and_extract` takes this dictionary and downloads the data files. Once the files are downloaded, :func:`datasets.SplitGenerator` organizes each split in the dataset. This is a simple class that contains:

Reviewer: ```suggestion
2. :func:`datasets.DownloadManager.download_and_extract` takes this dictionary and downloads the data files. Once the files are downloaded, use :class:`datasets.SplitGenerator` to organize each split in the dataset. This is a simple class that contains:
```

> * The :obj:`name` of each split. You should use the standard split names: :obj:`datasets.Split.TRAIN`, :obj:`datasets.Split.TEST`, and :obj:`datasets.Split.VALIDATION`.

Reviewer: ```suggestion
  * The :obj:`name` of each split. You should use the standard split names: :obj:`datasets.Split.TRAIN`, :obj:`datasets.Split.TEST`, and :obj:`datasets.Split.VALIDATION`.
```

> * :obj:`gen_kwargs` provides the file paths to the data files to load for each split.

Reviewer: ```suggestion
  * :obj:`gen_kwargs` provides the file paths to the data files to load for each split.
```

Your :obj:`datasets.DatasetBuilder._split_generator()` should look like this now:

.. code-block::

   def _split_generators(self, dl_manager: datasets.DownloadManager) -> List[datasets.SplitGenerator]:
       urls_to_download = self._URLS
       downloaded_files = dl_manager.download_and_extract(urls_to_download)

       return [
           datasets.SplitGenerator(name=datasets.Split.TRAIN, gen_kwargs={"filepath": downloaded_files["train"]}),
           datasets.SplitGenerator(name=datasets.Split.VALIDATION, gen_kwargs={"filepath": downloaded_files["dev"]}),
       ]

Generate samples
----------------

At this point, you have:

* Added the dataset attributes.
* Provided instructions for how to download the data files.
* Organized the splits.

The next step is to actually generate the samples in each split. 

1. :obj:`datasets.DatasetBuilder._generate_examples` takes the file path provided by :obj:`gen_kwargs` to read and parse the data files. You need to write a function that loads the data files and extracts the columns.

2. Your function should yield a tuple of an ``id_``, and an example from the dataset.

.. code-block::

>    def _generate_examples(self, filepath):
>    """This function returns the examples in the raw (text) form."""
>    logger.info("generating examples from = %s", filepath)
>    with open(filepath) as f:
>        squad = json.load(f)
>        for article in squad["data"]:
>            title = article.get("title", "").strip()
>            for paragraph in article["paragraphs"]:
>                context = paragraph["context"].strip()
>                for qa in paragraph["qas"]:
>                    question = qa["question"].strip()
>                    id_ = qa["id"]
> 
>                    answer_starts = [answer["answer_start"] for answer in qa["answers"]]
>                    answers = [answer["text"].strip() for answer in qa["answers"]]
> 
>                    # Features currently used are "context", "question", and "answers".
>                    # Others are extracted here for the ease of future expansions.
>                    yield id_, {
>                        "title": title,
>                        "context": context,
>                        "question": question,
>                        "id": id_,
>                        "answers": {"answer_start": answer_starts, "text": answers,},
>                 }

Reviewer: ```suggestion
   def _generate_examples(self, filepath):
       """This function returns the examples in the raw (text) form."""
       logger.info("generating examples from = %s", filepath)
       with open(filepath) as f:
           squad = json.load(f)
           for article in squad["data"]:
               title = article.get("title", "").strip()
               for paragraph in article["paragraphs"]:
                   context = paragraph["context"].strip()
                   for qa in paragraph["qas"]:
                       question = qa["question"].strip()
                       id_ = qa["id"]

                       answer_starts = [answer["answer_start"] for answer in qa["answers"]]
                       answers = [answer["text"].strip() for answer in qa["answers"]]

                       # Features currently used are "context", "question", and "answers".
                       # Others are extracted here for the ease of future expansions.
                       yield id_, {
                           "title": title,
                           "context": context,
                           "question": question,
                           "id": id_,
                           "answers": {"answer_start": answer_starts, "text": answers,},
                       }
```
